odc/geo/gridspec.py

```diff
@@ -22,46 +34,47 @@ class GridSpec:
     :param crs:
        Coordinate System used to define the grid
     :param tile_shape:
-       ``(Y, X)`` size of each tile in pixels
+       Size of each tile in pixels
     :param resolution:
-       ``(Y, X)`` size of each data point in the grid, in CRS units. ``Y`` will usually be negative
+       Size of each data point in the grid, in CRS units. ``Y`` will usually be negative.
     :param origin:
-       ``(X, Y)`` coordinate of a bottom-left corner of the ``(0,0)`` tile in CRS units. Default is
-       ``(0.0, 0.0)``
+       Coordinate of a bottom-left corner of the ``(0,0)`` tile in CRS units. Default is
+       ``xy_(0.0, 0.0)``
     :param flipx: when ``True`` grid index for X axis increments left to right
     :param flipy: when ``True`` grid index for Y axis increments top to bottom
     """
 
     def __init__(
         self,
         crs: SomeCRS,
-        tile_shape: Tuple[int, int],
-        resolution: Union[float, Tuple[float, float]],
-        origin: Optional[Tuple[float, float]] = None,
+        tile_shape: SomeShape,
+        resolution: SomeResolution,
+        origin: Optional[XY[float]] = None,
         flipx: bool = False,
         flipy: bool = False,
     ):
-        if isinstance(resolution, (int, float)):
-            resolution = (-resolution, resolution)
-        resolution = float(resolution[0]), float(resolution[1])
+        resolution = res_(resolution)
 
         if origin is None:
-            origin = (0.0, 0.0)
+            origin = xy_(0.0, 0.0)
         else:
-            origin = float(origin[0]), float(origin[1])
+            assert isinstance(origin, XY)
+
+        if isinstance(tile_shape, tuple):
+            tile_shape = yx_(tile_shape)
 
         self.crs = norm_crs_or_error(crs)
         self._shape = tile_shape
         self.resolution = resolution
-        self.tile_size = (
-            tile_shape[0] * abs(resolution[0]),
-            tile_shape[1] * abs(resolution[1]),
+        self.tile_size = xy_(
+            tile_shape.x * abs(resolution.x),
+            tile_shape.x * abs(resolution.y),
```

Comment:
```suggestion
            tile_shape.y * abs(resolution.y),
```

clearly tests are not checking non-square tiles